### PR TITLE
[wip] Added bulk add to DC & BO Assets.

### PR DESCRIPTION
**Pull request by ghost**
Added new field MultilineField which handles seperated values.
Added MultivalueFormMixin which validates many MultilineField in one form.

connected to: #1606

### CODE REVIEW

**Comment by vi4m** [comment]
screenshotplease.

File: src/ralph/admin/fields.py [hunk 1]
```diff
@@ -0,0 +1,109 @@
...
+                raise forms.ValidationError(_("There are duplicates in field."))
+
+    def to_python(self, value):
+        items = []
+        if value:
+            for item in re.split(self.separators, value):
+                items.append(item.strip())
+        return items
+
+    def clean(self, value):
```
Comment (mkurek) [inline src/ralph/admin/fields.py]: is this method necessary if it only call super?

File: src/ralph/admin/fields.py [hunk 2]
```diff
@@ -0,0 +1,109 @@
...
+        value = super(MultilineField, self).clean(value)
+        return value
+
+
+class MultivalueFormMixin(object):
+    """A form that has several multiline fields that need to have the
+    same number of entries.
+
+    :param multivalue_fields: list of form fields which require the same item
```
Comment (mkurek) [inline src/ralph/admin/fields.py:52]: describe `one_of_mulitvalue_required` as well

File: src/ralph/admin/fields.py [hunk 3]
```diff
@@ -0,0 +1,109 @@
...
+    by `separators`
+    Validation:
+        - separated values cannot duplicate each other,
+    """
+    widget = forms.Textarea
+    separators = ",|\n"
+
+    def __init__(self, allow_duplicates=True, *args, **kwargs):
+        self.allow_duplicates = allow_duplicates
+        super(MultilineField, self).__init__(*args, **kwargs)
```
Comment (mkurek) [inline src/ralph/admin/fields.py]: we should stick to one `super` convention - either `super(MultilineField, self).__init__(*args, **kwargs)` or `super().__init__(*args, **kwargs)`

File: src/ralph/admin/fields.py [hunk 4]
```diff
@@ -0,0 +1,109 @@
...
+    :param multivalue_fields: list of form fields which require the same item
+    count
+    """
+    multivalue_fields = []
+    one_of_mulitvalue_required = []
+
+    @transaction.atomic
+    def save_model(self, request, obj, form, change):
+        super().save_model(request, obj, form, change)
+        obj = self.model(**form.cleaned_data)
```
Comment (mkurek) [inline src/ralph/admin/fields.py]: why don't you use `obj` passed to this method?

File: src/ralph/admin/fields.py [hunk 5]
```diff
@@ -0,0 +1,109 @@
...
+        for idx, form_values in enumerate(self.multivalues_rows()):
+            for value, field_name in zip(form_values, self.multivalue_fields):
+                setattr(obj, field_name, value)
```
Comment (mkurek) [inline src/ralph/admin/fields.py]: shouldn't `super().save_model(request, obj, form, change)` be called there?

File: src/ralph/back_office/admin.py [hunk 6]
```diff
@@ -35,6 +48,7 @@ class BackOfficeAssetAdmin(
     list_select_related = ['model', 'user', 'warehouse', 'model__manufacturer']
     raw_id_fields = ['model', 'user', 'owner', 'service_env']
     resource_class = resources.BackOfficeAssetResource
+    form = BackOfficeAssetForm
```
Comment (mkurek) [inline src/ralph/back_office/admin.py:55]: this form should be used only to add asset - not to edit

File: src/ralph/admin/fields.py [hunk 7]
```diff
@@ -0,0 +1,109 @@
...
+
+
+class MultivalueFormMixin(object):
+    """A form that has several multiline fields that need to have the
+    same number of entries.
+
+    :param multivalue_fields: list of form fields which require the same item
+    count
+    """
+    multivalue_fields = []
```
Comment (mkurek) [inline src/ralph/admin/fields.py:64]: wouldn't it be better to define it in form's meta? @ar4s what do you think?

File: src/ralph/admin/fields.py [hunk 8]
```diff
@@ -0,0 +1,109 @@
...
+        super().save_model(request, obj, form, change)
+        obj = self.model(**form.cleaned_data)
+        for idx, form_values in enumerate(self.multivalues_rows()):
+            for value, field_name in zip(form_values, self.multivalue_fields):
+                setattr(obj, field_name, value)
+        msg = _('Successfully added {} assets'.format(idx + 1))
+        self.message_user(request, msg)
+
+    def equal_count_validator(self, cleaned_data):
+        """Adds a validation error if if form's multivalues fields have
```
Comment (mkurek) [inline src/ralph/admin/fields.py]: duplicated `if`

File: src/ralph/admin/tests/tests_fields.py [hunk 9]
```diff
@@ -0,0 +1,90 @@
...
+            'sn': 'sn1,,sn3',
+            'barcode': 'br1,,br3',
+        }
+        form = OneRequiredTestForm(data)
+        self.assertFalse(form.is_valid())
+        self.assertIn('sn', form.errors)
+        self.assertIn('barcode', form.errors)
+
+
+class MultilineFieldTest(SimpleTestCase):
```
Comment (mkurek) [inline src/ralph/admin/tests/tests_fields.py:94]: could you write additional test when none of required fields is filled (fields have blank values)?

File: src/ralph/back_office/admin.py (hunk 6 above)
Comment (ghost) [inline src/ralph/back_office/admin.py:55]: done https://github.com/allegro/ralph/pull/1679/files#diff-499a76f5c9c2f4f033091d2293326317R25

File: src/ralph/admin/tests/tests_functional.py [hunk 10]
```diff
@@ -1,4 +1,4 @@
-# -*- coding: utf-8 -*-
+# -*- cmding: utf-8 -*-
```
Comment (ar4s) [inline src/ralph/admin/tests/tests_functional.py:1]: wut?

File: src/ralph/back_office/admin.py [hunk 11]
```diff
@@ -12,6 +14,21 @@
...
+        model = BackOfficeAsset
+        fields = '__all__'
+
+    def __init__(self, *args, **kwargs):
+        super().__init__(*args, **kwargs)
+        if not self.instance.id:
+            self.fields['sn'] = MultilineField('sn')
+            self.fields['barcode'] = MultilineField('barcode')
+            self.multivalue_fields = ['sn', 'barcode']
+            self.one_of_mulitvalue_required = ['sn', 'barcode']
```
Comment (ar4s) [inline src/ralph/back_office/admin.py:29]: `self.Meta.multivalue_fields = ['sn', 'barcode']`

**Pull request closed without merging**